### PR TITLE
Remove support for zeek-4.x.

### LINKED ISSUE #153: Remove support for Zeek 4.x.
I'd say time has come to limit the plugin to Zeek 5.x.

### PR DESCRIPTION
**Pull request by bbannier**
Closes #153.

### CODE REVIEW

File: .cirrus.yml [hunk 1]
```diff
@@ -14,12 +14,6 @@ zkg_task:
   container:
     dockerfile: ci/Dockerfile
     matrix:
-      - docker_arguments:
-        - ZEEK_VERSION: "v4.0.3"
-        - ZEEK_CONFFLAGS: ""
-      - docker_arguments:
-        - ZEEK_VERSION: "v4.2.1"
-        - ZEEK_CONFFLAGS: ""
```
Comment (rsmmr) [inline .cirrus.yml:22]: Let's add the most recent 5.x version.

File: src/plugin.cc [hunk 2]
```diff
@@ -247,7 +247,7 @@ void plugin::Zeek_Spicy::Plugin::registerFileAnalyzer(const std::string& name,
     // this point already, so ours won't get initialized anymore.
     c->Initialize();
 
-    trackComponent(c, c->Tag().Type()); // must come after Initialize(); needed for Zeek < 4.2
```
Comment (rsmmr) [inline src/plugin.cc:250]: Not sure what this was about, but the comment seems to say something's not needed anymore now?

File: src/plugin.cc [hunk 3]
```diff
@@ -299,7 +299,7 @@ void plugin::Zeek_Spicy::Plugin::registerPacketAnalyzer(const std::string& name,
...
+    trackComponent(c, c->Tag().Type()); // Must come after Initialize().
```
Comment (rsmmr) [inline src/plugin.cc:302]: Same.

File: include/zeek-compat.h [hunk 4]
```diff
@@ -55,24 +41,21 @@
...
 #include <zeek/analyzer/Analyzer.h>
 #include <zeek/analyzer/Manager.h>
 #include <zeek/analyzer/protocol/pia/PIA.h>
 #include <zeek/analyzer/protocol/tcp/TCP.h>
-#if ZEEK_VERSION_NUMBER < 40100 // Zeek < 4.1
-#include <zeek/Sessions.h>
-#include <zeek/analyzer/protocol/udp/UDP.h>
-#else // Zeek >= 4.1
-#include <zeek/session/Manager.h>
-#endif
```
Comment (rsmmr) [inline include/zeek-compat.h:70]: From the diff it's hard to say if some of the `#else` blocks being removed were triggering for Zeek >= 5.0. Please just double-check that nothing goes out accidentally (although then things would presumably break, so probably fine as it is).

**Review by rsmmr — APPROVED** [review]
One more thing to check, but hard to see from the diff: If there are now cases in `zeek::compat` where we don't have to version-switch at all anymore, just update the calling code to remove the usage of the `compat` wrapper altogether.

File: include/zeek-compat.h (hunk 4 above)
Comment (bbannier) [reply include/zeek-compat.h:70]: This should be fine, the include for `zeek/session/Manager.h` stays, the one for `zeek/Sessions.h` is removed. What makes these include changes harder to review is that they also end up resorted (previously the preprocessor calls acted as barriers).

File: src/plugin.cc (hunk 2 above)
Comment (bbannier) [reply src/plugin.cc:250]: These `trackComponent` calls update `Plugin::_analyzer_name_to_tag_type` which we seem to be using to update linker scopes for pre-registered analyzers, see e.g., https://github.com/zeek/spicy-plugin/blob/09184f9f707b275c90da50a07ca7466dfc7ecd4c/src/plugin.cc#L163-L178. Not sure, but maybe we'll be able to remove this once we remove plugin JIT mode.

File: .cirrus.yml (hunk 1 above)
Comment (bbannier) [reply .cirrus.yml:22]: Not sure I am missing something, but just below we run against `v5.1.1` which should be the latest release.

File: src/plugin.cc (hunk 3 above)
Comment (bbannier) [reply src/plugin.cc:302]: See above.

**Comment by bbannier** [comment]
> One more thing to check, but hard to see from the diff: If there are now cases in `zeek::compat` where we don't have to version-switch at all anymore, just update the calling code to remove the usage of the `compat` wrapper altogether.

I went through the namespace and nothing in it remains. I added two new patches to the PR cleaning it up, PTAL.

File: .cirrus.yml (hunk 1 above)
Comment (rsmmr) [reply .cirrus.yml:22]: I'm just blind. :-)

**Review by rsmmr — APPROVED** [review]
Great to see the new, tiny `zeek-compat.h`!

**Pull request merged**